### ISSUE TITLE
Confusing behavior with valid-jsdoc and requireReturn

### ISSUE THREAD

**Issue by tschaub** [eslint/eslint#4859]
While looking into a way to make the `valid-jsdoc` rule work well with `@interface` (where functions may be required to have no body), I noticed some odd behavior with `requireReturn`.

Using ESLint @ 009a949ccbcd3c7ed9964a4583d7814a398ecfd4, the following is valid:
```js
/* eslint valid-jsdoc: [2, {"requireReturn": true}] */

/**
 * Testing valid-jsdoc.
 * @return {string} A string.
 */
exports.foo = function() {
  // noop
};
```

If I change `requireReturn` to `false`, I get an error:
```js
/* eslint valid-jsdoc: [2, {"requireReturn": false}] */

/**
 * Testing valid-jsdoc.
 * @return {string} A string.
 */
exports.foo = function() {
  // noop
};
```

> Unexpected @return tag; function has no return statement  valid-jsdoc

So only if I don't require a `@return` tag, the function is checked for a return statement.  Here is the [relevant code](https://github.com/eslint/eslint/blob/009a949ccbcd3c7ed9964a4583d7814a398ecfd4/lib/rules/valid-jsdoc.js#L145-L147):
```js
if (!requireReturn && !functionData.returnPresent && (tag.type === null || !isValidReturnType(tag))) {
  context.report(jsdocNode, "Unexpected @" + tag.title + " tag; function has no return statement.");
}
```

The `!requireReturn` looks like a typo.  It would make a bit more sense to say "if I don't require the return tag, don't check if a return statement is present" - but even that may be overloading `requireReturn` (maybe the option should be about the tags only and not about the code).

I'm also confused by this sentence in [the documentation](https://github.com/eslint/eslint/blob/009a949ccbcd3c7ed9964a4583d7814a398ecfd4/docs/rules/valid-jsdoc.md#requirereturn):

> Note that with this option set to false, if there is a return in the function, an error will still be logged and if there is a @return specified and there are no return statements in the function an error will also be logged.

Specifically, this is confusing: "if there is a return in the function, an error will still be logged *and* (??) if [...] *there are no return statements in the function* [...]".  It is not clear to me how there can be a "return in the function" and "no return statements in the function."

@briandipalma and @nzakas - can you confirm this is the behavior you intended to get out of #673?

**Comment by eslintbot** [comment]
@tschaub Thanks for the issue! If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

**Comment by nzakas** [comment]
This is expected behavior. When `requireReturn` is false, it means that you should not have `@returns` in your JSDoc unless there is a `return` in the function. In your example, you get an error because you are including `@return` and there is no `return` in the function. 

We could make the docs a bit clearer, but unless I'm missing something, I don't see a bug here.

**Comment by tschaub** [comment]
It makes sense to me that if I have a `return` statement and no `@return` tag, `valid-jsdoc` would complain.

What is confusing to me is that to disable that behavior, I would set `requireReturn` to `true`.

**Comment by tschaub** [comment]
Or, put another way, if I want `valid-jsdoc` to fail when there is no `@return` tag for a function that has a `return` statement, it is confusing to me that I would set `requireReturn` to `false`.  The confusing part is that the rule becomes more strict when I say "don't require a return".

**Comment by nzakas** [comment]
The way the rule is designed, all the flags are enabled by default and you can opt-out of the ones you don't want. We had to do this for backwards compatibility as we added more features. 

I think you're misunderstanding how this options works. `requireReturn: true` (the default) says "You must always use `@returns` no matter what." `requireReturn: false` says, "You must use `@returns` only if you have a return value in your function."

So by default, you will get an error if you omit `@returns` from a function's JSDoc regardless of what's going on in the function. You can opt-out of that behavior by setting `requireReturn: false`. There's never a need to manually set `requireReturn: true` because that's the default.

**Comment by tschaub** [comment]
> requireReturn: true (the default) says "You must always use @returns no matter what." requireReturn: false says, "You must use @returns only if you have a return value in your function."

That makes sense.

But the code does more than that.  See my example from the top.

```js
/**
 * The docs say this function returns a string.
 * @returns {string} Some string.
 */
function noop() {}
```

Why shouldn't `valid-jsdoc` fail by default above?  It only fails if `requireReturn` is `false`.

**Comment by tschaub** [comment]
Ok, thanks for your patience @nzakas.  I've made peace with `requireReturn`.

#4864 is my attempt to clarify the docs.

While it would be a breaking change, I think it would be unsurprising if (by default) any function with a `return` statement must be documented with `@return`, and any function with a `@return` tag must include a `return` statement.  And to disable this "consistent return" checking `requireReturn` could be set to `false` (or maybe it would be called `consistentReturn`).

**Comment by Telokis** [comment]
@nzakas Is there a way to prevent the rule from generating a warning/error if there is a `return` statement that is not documented?
In asynchronous code `return` is only used to prevent the code from executing the end of the function. (For example, by doing `return callback(err);`.
So I'd like to **not** enforce the presence of `@returns` in my doc. That's what I expected from `requireReturn`.

**Comment by briandipalma** [comment]
This wouldn't be `requireReturn`'s use case. In essence you want to enforce documentation *apart* from return statements? Or do you want to disable enforcing return documentation on statements that consist of the pattern `callback(args);`?

**Comment by Telokis** [comment]
@briandela Precisely, I'd like to disable `@returns` necessity when the `return` line matches a specific pattern. Would it be possible with the current eslint?

**Comment by briandipalma** [comment]
I'm sure it would be but I'd raise a feature request instead of commenting on a closed issue if that's what you want. Try and implement the feature.

**Comment by briandipalma** [comment]
Just take a look at the JSDoc rule and add a new param for the feature you want.
